content/unroll-default-arguments.md

```diff
@@ -0,0 +1,797 @@
+   stays.
+
+### Challenges of Non-Final Methods and Overriding
+
+To elaborate a bit on the issues with non-final methods and overriding, consider the following 
+case with four classes, `Upstream`, `Downstream`, `Main1` and `Main2`, each of which is compiled
+against different versions of each other (hence the varying number of parameters for `foo`):
+
+```scala
+class Upstream{ // V2
+   def foo(s: String, n: Int = 1, @unroll b: Boolean = true, @unroll l: Long = 0) = s + n + b + l
+}
+```
+
+```scala
+class Downstream extends Upstream{ // compiled against Upstream V1
+   override def foo(s: String, n: Int = 1) = super.foo(s, n) + s + n
+}
+```
+
+```scala
+object Main1 { // compiled against Upstream V2
+   def main(args: Array[String]): Unit = {
+      new Downstream().foo("hello", 123, false, 456L)
+   }
+}
+```
+
+```scala
+object Main2 { // compiled against Upstream V1
+   def main(args: Array[String]): Unit = {
+      new Downstream().foo("hello", 123)
+   }
+}
+```
+
+
+The challenge here is: how do we make sure that `Main1` and `Main2`, who call
+`new Downstream().foo`, correctly pick up the version of `def foo` that is 
+provided by `Downstream`? 
+
+With the current implementation, the `override def foo` inside `Downstream` would only
+override one of `Upstream`'s synthetic forwarders, but would not override the actual
+primary implementation. As a result, we would see `Main1` calling the implementation
+of `foo` from `Upstream`, while `Main2` calls the implementation of `foo` from 
+`Downstream`. So even though both `Main1` and `Main2` have the same 
+`Upstream` and `Downstream` code on the classpath, they end up calling different
+implementations based on what they were compiled against.
+
+We cannot perform the method search and dispatch _within_ the `def foo` methods,
+because it depends on exactly _how_ `foo` is called: the `InvokeVirtual` call from
+`Main1` is meant to resolve to `Downstream#foo`, while the `InvokeSpecial` call
+from `Downstream#foo`'s `super.foo` is meant to resolve to `Upstream#foo`. But a
+method implementation cannot know how it was called, and thus it is impossible
+for `def foo` to forward the call to the right place.
+
+This issue only arises in the presence of version skew between `Upstream` and 
+`Downstream` code, but that scenario is precisely where `@unroll` is meant to
+provide benefits. Thus, unless we can find some solution, we cannot properly support
+virtual methods and overrides in `@unroll`.
+
+It may be possible to loosen this restriction to also allow abstract methods that
+are implemented only once by a final method. See the section about 
+[Abstract Methods](#abstract-methods) for details.
+
+## Major Alternatives
+
+The major alternatives to `@unroll` are listed below:
+
+1. [data-class](https://index.scala-lang.org/alexarchambault/data-class)
+2. [SBT Datatype](https://www.scala-sbt.org/1.x/docs/Datatype.html)
```

Comment:
Could you link to the current incarnation of sbt-datatype, which is [Contraband](https://www.scala-sbt.org/contraband/) instead please? Contraband primarily uses [GraphQL Schema](https://www.scala-sbt.org/contraband/schema.html) as the schema language with a few extensions like using `@since("1.0.0")` to denote when a field was added, though we do support the JSON-style schema for internal usage.

In addition to being binary-JAR compatible, Contraband supports generation of serialization codec, which is an important aspect similar to other schemas like Protobuf and Avro. Ideally, we should support cross talk between two services with all combinations of v1 -> v2, v2 -> v1, etc.

Reply:
Thanks! I've updated the link to contraband.